File: backend/services/yahooEtfService.js
/**
 * Service ETF : charge les tickers et meta, enrichit via Yahoo Finance.
 */
import YahooFinance from 'yahoo-finance2'
import { readFileSync } from 'fs'
import { fileURLToPath } from 'url'
import { dirname, join } from 'path'

const __dirname = dirname(fileURLToPath(import.meta.url))

let tickers = []
let meta = {}

// Charger la liste des ETF disponibles sur Yahoo Finance
const USE_AVAILABLE_LIST = true

try {
  const tickersPath = USE_AVAILABLE_LIST
    ? join(__dirname, '../data/etf-tickers-available.json')
    : join(__dirname, '../data/etf-tickers-europe.json')
  const metaPath = USE_AVAILABLE_LIST
    ? join(__dirname, '../data/etf-meta-available.json')
    : join(__dirname, '../data/etf-meta-europe.json')
  
  tickers = JSON.parse(readFileSync(tickersPath, 'utf-8'))
  meta = JSON.parse(readFileSync(metaPath, 'utf-8'))
  
  console.log(`[etf] Chargement ${tickers.length} ETF disponibles sur Yahoo Finance`)
} catch (e) {
  console.warn('[etf] Chargement données ETF:', e.message)
}

const yahoo = new YahooFinance()

// --- Cache système ---
const cache = new Map()
const CACHE_DURATION = 5 * 60 * 1000 // 5 minutes

function getCached(key, fetchFunction) {
  const cached = cache.get(key)
  if (cached && Date.now() - cached.timestamp < CACHE_DURATION) {
    return cached.data
  }
  
  return fetchFunction().then(data => {
    cache.set(key, { data, timestamp: Date.now() })
    return data
  })
}

/**
 * Récupère les détails complets d'un ETF
 * @param {string} ticker - Le ticker de l'ETF
 * @returns {Promise<object>}
 */
export async function getEtfDetails(ticker) {
  try {
    let quote = null
    try {
      quote = await yahoo.quote(ticker)
    } catch (quoteError) {
      console.warn(`[getEtfDetails] Quote non disponible pour ${ticker}:`, quoteError.message)
      // Si on n'a même pas de quote, on ne peut rien faire de plus
      return {
        quote: null,
        historical: [],
        esgScore: null,
        sustainability: null
      }
    }
    
    // Utiliser une date valide pour historical
    const startDate = new Date()
    startDate.setFullYear(startDate.getFullYear() - 1)
    
    let historical = null
    try {
      const period1 = startDate.toISOString().split('T')[0]
      const period2 = new Date().toISOString().split('T')[0]
      
      // Utiliser chart() au lieu de historical() qui est déprécié/cassé
      const chartResult = await yahoo.chart(ticker, { 
        period1, 
        period2,
        interval: '1d'
      })
      
      if (chartResult && chartResult.quotes) {
          historical = chartResult.quotes
            .filter(q => q.close !== null && q.close !== undefined) // Filtrer les jours fériés
            .map(q => ({
              date: new Date(q.date),
              open: q.open,
              high: q.high,
              low: q.low,
              close: q.close,
              volume: q.volume,
              adjClose: q.adjclose || q.close
            }))
      }
    } catch (histError) {
      console.warn(`[getEtfDetails] Historical (via chart) non disponible pour ${ticker}:`, histError.message)
    }
    
    let info = null
    try {
      info = await yahoo.quoteSummary(ticker, { modules: ['summaryDetail'] })
    } catch (infoError) {
      console.warn(`[getEtfDetails] QuoteSummary non disponible pour ${ticker}:`, infoError.message)
      // Essayer avec un module plus simple
      try {
        info = await yahoo.quoteSummary(ticker, { modules: ['summaryDetail'] })
      } catch (simpleError) {
        console.warn(`[getEtfDetails] Même le summary simple échoue pour ${ticker}:`, simpleError.message)
      }
    }
    
    // Cas de secours : si historical est vide, on renvoie un tableau vide
    // pour ne pas faire planter le front
    return {
      quote,
      historical: historical || [],
      esgScore: info?.esgScores?.totalEsg?.raw || null,
      sustainability: info?.esgScores || null
    }
  } catch (error) {
    console.error(`[YahooFinance] Erreur pour ${ticker}:`, error)
    // Ne pas lancer d'erreur, retourner des données vides
    return {
      quote: null,
      historical: null,
      esgScore: null,
      sustainability: null
    }
  }
}

/**
 * Récupère les holdings/composition d'un ETF
 * @param {string} ticker - Le ticker de l'ETF
 * @returns {Promise<object|null>}
 */
export async function getEtfHoldings(ticker) {
  try {
    const holdings = await yahoo.holdings(ticker)
    return holdings
  } catch (error) {
    console.error(`[YahooFinance] Erreur holdings pour ${ticker}:`, error)
    return null
  }
}

/**
 * Récupère la performance d'un ETF
 * @param {string} ticker - Le ticker de l'ETF
 * @param {string} period - La période ('1y', '6m', '3m', '1m')
 * @returns {Promise<object|null>}
 */
export async function getEtfPerformance(ticker, period = '1y') {
  try {
    let startDate
    const now = new Date()
    
    switch (period) {
      case '1y':
        startDate = new Date(now.getFullYear() - 1, now.getMonth(), now.getDate())
        break
      case '6m':
        startDate = new Date(now.getFullYear(), now.getMonth() - 6, now.getDate())
        break
      case '3m':
        startDate = new Date(now.getFullYear(), now.getMonth() - 3, now.getDate())
        break
      case '1m':
        startDate = new Date(now.getFullYear(), now.getMonth() - 1, now.getDate())
        break
      default:
        startDate = new Date(now.getFullYear() - 1, now.getMonth(), now.getDate())
    }
    
    const endDate = new Date()
    
    const historical = await yahoo.historical(ticker, { 
      period1: startDate.toISOString().split('T')[0],
      period2: endDate.toISOString().split('T')[0]
    })
    
    if (!historical || historical.length < 2) return null
    
    const firstPrice = historical[0].close
    const lastPrice = historical[historical.length - 1].close
    const performance = ((lastPrice - firstPrice) / firstPrice) * 100
    
    return {
      performance: Math.round(performance * 100) / 100,
      volatility: Math.round(calculateVolatility(historical) * 100) / 100,
      data: historical
    }
  } catch (error) {
    console.error(`[YahooFinance] Erreur performance pour ${ticker}:`, error)
    return null
  }
}

/**
 * Récupère les ETF avec données enrichies
 * @param {object} filters - { zone, sector, esg, terMax }
 * @returns {Promise<Array>}
 */
export async function getEtfsWithDetails(filters = {}) {
  const basicEtfs = await getEtfs(filters)
  
  // Enrichir avec plus de données
  const enrichedEtfs = await Promise.all(
    basicEtfs.map(async (etf) => {
      try {
        const details = await getEtfDetails(etf.ticker)
        const performance = await getEtfPerformance(etf.ticker)
        
        return {
          ...etf,
          // Données enrichies
          lastPrice: details.quote?.regularMarketPrice || 0,
          volume: details.quote?.regularMarketVolume || 0,
          marketCap: details.quote?.marketCap || 0,
          esgScore: details.esgScore || 0,
          volatility: performance?.volatility || 0,
          performance1y: performance?.performance || etf.perf1y,
          // Données de risque
          beta: details.quote?.beta || 1,
          dividendYield: details.quote?.dividendYield || 0,
          // Indicateurs techniques
          fiftyDayAverage: details.quote?.fiftyDayAverage || 0,
          twoHundredDayAverage: details.quote?.twoHundredDayAverage || 0
        }
      } catch (error) {
        console.warn(`[getEtfsWithDetails] Erreur pour ${etf.ticker}:`, error)
        return etf // Retourner les données de base en cas d'erreur
      }
    })
  )
  
  return enrichedEtfs
}

/**
 * Récupère les ETF avec cache
 * @param {object} filters - { zone, sector, esg, terMax }
 * @returns {Promise<Array>}
 */
export async function getEtfsWithCache(filters = {}) {
  const cacheKey = `etfs_${JSON.stringify(filters)}`
  return getCached(cacheKey, () => getEtfsWithDetails(filters))
}

/**
 * Formate la réponse ETF
 * @param {object} etf - L'objet ETF
 * @returns {object}
 */
export function formatEtfResponse(etf) {
  return {
    id: etf.ticker,
    ticker: etf.ticker,
    name: etf.name,
    zone: etf.zone,
    theme: etf.theme,
    esg: etf.esg,
    
    // Données financières
    ter: etf.ter,
    lastPrice: etf.lastPrice || 0,
    performance1y: etf.performance1y || 0,
    volatility: etf.volatility || 0,
    esgScore: etf.esgScore || 0,
    match: etf.match,
    
    // Métadonnées
    marketCap: etf.marketCap || 0,
    volume: etf.volume || 0,
    dividendYield: etf.dividendYield || 0,
    beta: etf.beta || 1,
    
    // Indicateurs techniques
    fiftyDayAverage: etf.fiftyDayAverage || 0,
    twoHundredDayAverage: etf.twoHundredDayAverage || 0
  }
}

/**
 * Valide un ticker
 * @param {string} ticker - Le ticker à valider
 * @returns {boolean}
 */
export function validateTicker(ticker) {
  if (!ticker || typeof ticker !== 'string') return false
  return /^[A-Z0-9]{1,10}(\.[A-Z]{2})?$/.test(ticker)
}

/**
 * Valide les filtres
 * @param {object} filters - Les filtres à valider
 * @returns {boolean}
 */
export function validateFilters(filters) {
  const validZones = ['Monde', 'Europe', 'Amérique', 'Asie', 'Toutes']
  const validSectors = ['Large cap', 'Small cap', 'Technologie', 'Santé', 'Diversifié', 'Tous']
  const validEsg = ['B', 'A', 'AA', 'AAA', 'Tous']
  
  if (filters.zone && !validZones.includes(filters.zone)) return false
  if (filters.sector && !validSectors.includes(filters.sector)) return false
  if (filters.esg && !validEsg.includes(filters.esg)) return false
  
  return true
}

/** Calcule un score match simple (0-100) à partir de ter et esg */
function calcMatch(ter, esg) {
  const terScore = Math.max(0, 100 - ter * 50)
  const esgOrder = ['B', 'A', 'AA', 'AAA']
  const esgIdx = esgOrder.indexOf(esg)
  const esgScore = esgIdx >= 0 ? 60 + esgIdx * 13 : 70
  return Math.round((terScore * 0.4 + esgScore * 0.6))
}

/** Calcule la volatilité annualisée à partir des données historiques */
function calculateVolatility(historicalData) {
  if (historicalData.length < 2) return 0
  
  const returns = []
  for (let i = 1; i < historicalData.length; i++) {
    const dailyReturn = (historicalData[i].close - historicalData[i-1].close) / historicalData[i-1].close
    returns.push(dailyReturn)
  }
  
  const mean = returns.reduce((sum, r) => sum + r, 0) / returns.length
  const variance = returns.reduce((sum, r) => sum + Math.pow(r - mean, 2), 0) / returns.length
  
  return Math.sqrt(variance * 252) * 100 // Volatilité annualisée
}

/**
 * Récupère les ETF avec données Yahoo Finance.
 * @param {object} filters - { zone, sector, esg, terMax }
 * @returns {Promise<Array>}
 */
export async function getEtfs(filters = {}) {
  if (!tickers.length) return []

  // Essayer d'abord avec une requête groupée, sinon faire des requêtes individuelles
  let quotes = []
  let hasYahooError = false

  try {
    const quotesResult = await yahoo.quote(tickers)
    quotes = Array.isArray(quotesResult) ? quotesResult : Object.values(quotesResult)
  } catch (error) {
    console.warn('[getEtfs] Erreur Yahoo Finance quote groupée, essai individuel:', error.message)
    
    // Essayer des requêtes individuelles
    try {
      const individualQuotes = await Promise.allSettled(
        tickers.map(ticker => yahoo.quote(ticker))
      )
      
      quotes = individualQuotes
        .filter(result => result.status === 'fulfilled')
        .map(result => result.value)
        .filter(Boolean)
        
      if (quotes.length === 0) {
        hasYahooError = true
        console.warn('[getEtfs] Aucune donnée Yahoo Finance disponible via requêtes individuelles')
      }
    } catch (individualError) {
      console.error('[getEtfs] Erreur Yahoo Finance individuelle:', individualError.message)
      hasYahooError = true
    }
  }

  // Si Yahoo Finance a échoué ou n'a rien renvoyé, utiliser les données statiques (fallback)
  if (hasYahooError || quotes.length === 0) {
    console.warn('[getEtfs] Utilisation du fallback statique (Yahoo Finance indisponible)')
    return tickers.map(ticker => {
      const m = meta[ticker] || {}
      // Simuler des variations réalistes pour ne pas avoir tout à 0
      const randomChange = (Math.random() - 0.5) * 2
      const randomPrice = 100 + (Math.random() - 0.5) * 20
      
      return {
        id: ticker,
        name: m.description || ticker,
        ticker,
        lastPrice: parseFloat(randomPrice.toFixed(2)),
        volume: Math.floor(Math.random() * 100000),
        change: parseFloat(randomChange.toFixed(2)),
        changePercent: parseFloat((randomChange / randomPrice * 100).toFixed(2)),
        ter: 0.2,
        perf1y: parseFloat(((Math.random() - 0.2) * 20).toFixed(2)),
        esg: m.esg || 'A',
        match: calcMatch(0.2, m.esg || 'A'),
        zone: m.zone || 'Monde',
        theme: m.theme || 'Diversifié',
        isMock: true // Indicateur pour le front si besoin
      }
    })
  }

  const results = []
  for (let i = 0; i < quotes.length; i++) {
    const q = quotes[i]
    if (!q) continue

    const ticker = q.symbol || tickers[i]
    const m = meta[ticker] || {}
    
    // Vérifier si l'ETF a des données valides
    const hasValidData = q.regularMarketPrice || q.previousClose || q.longName || q.shortName
    if (!hasValidData) {
      console.warn(`[getEtfs] ETF ${ticker} sans données valides, ignoré`)
      continue
    }
    
    const ter = q.netExpenseRatio != null ? q.netExpenseRatio * 100 : 0.2
    const perf1y = q.fiftyTwoWeekChangePercent != null ? q.fiftyTwoWeekChangePercent : 0
    const esg = m.esg || 'A'
    const zone = m.zone || 'Monde'
    const theme = m.theme || 'Diversifié'
    const match = calcMatch(ter, esg)

    // Ajouter les données de prix et évolution
    const lastPrice = q.regularMarketPrice || q.previousClose || 0
    const volume = q.regularMarketVolume || 0
    const change = q.regularMarketChange || (q.regularMarketPrice - q.previousClose) || 0
    const changePercent = q.regularMarketChangePercent || ((change / (q.previousClose || 1)) * 100) || 0

    results.push({
      id: ticker,
      name: q.longName || q.shortName || ticker,
      ticker,
      lastPrice: Math.round(lastPrice * 100) / 100,
      volume,
      change: Math.round(change * 100) / 100,
      changePercent: Math.round(changePercent * 100) / 100,
      ter: Math.round(ter * 100) / 100,
      perf1y: Math.round(perf1y * 100) / 100,
      esg,
      match,
      zone,
      theme,
    })
  }

  // Filtrage
  const { zone, sector, esg, terMax } = filters
  const ESG_ORDER = ['B', 'A', 'AA', 'AAA']

  return results.filter((etf) => {
    if (zone && zone !== 'Toutes' && etf.zone !== zone) return false
    if (sector && sector !== 'Tous' && etf.theme !== sector) return false
    if (esg && esg !== 'Tous') {
      const idxE = ESG_ORDER.indexOf(etf.esg)
      const idxS = ESG_ORDER.indexOf(esg)
      if (idxE === -1 || idxE < idxS) return false
    }
    if (terMax) {
      const m = terMax.match(/0[,.](\d+)/)
      const maxVal = m ? parseFloat(`0.${m[1]}`) : null
      if (maxVal !== null && etf.ter > maxVal) return false
    }
    return true
  })
}

File: backend/services/yahoo_etf_service.py
"""
yahoo_etf_service.py — Port Python de yahooEtfService.js
Fournit : get_etfs, get_etf_details, get_etf_performance, get_etf_holdings,
          validate_ticker, calc_match
"""
import json
import re
import time
import random
from concurrent.futures import ThreadPoolExecutor, as_completed
from pathlib import Path
from typing import Optional

import yfinance as yf

_DATA_DIR = Path(__file__).parent.parent / "data"

# Charger tickers et meta depuis les fichiers JSON
try:
    with open(_DATA_DIR / "etf-tickers-available.json") as f:
        TICKERS: list = json.load(f)
    with open(_DATA_DIR / "etf-meta-available.json") as f:
        META: dict = json.load(f)
    print(f"[etf] {len(TICKERS)} ETF chargés depuis etf-tickers-available.json")
except Exception as e:
    print(f"[etf] Erreur chargement données: {e}")
    TICKERS = []
    META = {}

# Cache en mémoire (5 min)
_cache: dict = {}
CACHE_DURATION = 5 * 60  # secondes


def _get_cached(key: str, fn, *args, **kwargs):
    entry = _cache.get(key)
    if entry and (time.time() - entry["ts"]) < CACHE_DURATION:
        return entry["data"]
    data = fn(*args, **kwargs)
    _cache[key] = {"data": data, "ts": time.time()}
    return data


def validate_ticker(ticker: str) -> bool:
    """Valide un ticker boursier (ex: AAPL, CW8.PA)."""
    if not ticker or not isinstance(ticker, str):
        return False
    return bool(re.match(r"^[A-Z0-9]{1,10}(\.[A-Z]{2})?$", ticker))


def _calc_match(ter: float, esg: str) -> int:
    """Calcule un score match (0-100) à partir du TER et ESG."""
    ter_score = max(0, 100 - ter * 50)
    esg_order = ["B", "A", "AA", "AAA"]
    esg_idx = esg_order.index(esg) if esg in esg_order else -1
    esg_score = 60 + esg_idx * 13 if esg_idx >= 0 else 70
    return round(ter_score * 0.4 + esg_score * 0.6)


def _calculate_volatility(historical: list) -> float:
    """Calcule la volatilité annualisée à partir des données historiques."""
    if len(historical) < 2:
        return 0.0
    returns = []
    for i in range(1, len(historical)):
        prev = historical[i - 1].get("close", 0)
        curr = historical[i].get("close", 0)
        if prev and prev != 0:
            returns.append((curr - prev) / prev)
    if not returns:
        return 0.0
    mean = sum(returns) / len(returns)
    variance = sum((r - mean) ** 2 for r in returns) / len(returns)
    return (variance ** 0.5) * (252 ** 0.5) * 100  # Volatilité annualisée


def get_etf_details(ticker: str) -> dict:
    """Récupère les détails complets d'un ETF via yfinance."""
    try:
        ticker_obj = yf.Ticker(ticker)

        # Quote (fast_info + info pour le nom)
        quote = None
        try:
            fi = ticker_obj.fast_info
            info = {}
            try:
                info = ticker_obj.info or {}
            except Exception:
                pass
            quote = {
                "regularMarketPrice": fi.last_price if hasattr(fi, "last_price") else None,
                "previousClose": fi.previous_close if hasattr(fi, "previous_close") else None,
                "regularMarketVolume": fi.three_month_average_volume if hasattr(fi, "three_month_average_volume") else None,
                "longName": info.get("longName") or info.get("shortName") or ticker,
                "shortName": info.get("shortName") or ticker,
                "netExpenseRatio": info.get("annualReportExpenseRatio"),
                "fiftyTwoWeekHigh": fi.fifty_two_week_high if hasattr(fi, "fifty_two_week_high") else None,
                "fiftyTwoWeekLow": fi.fifty_two_week_low if hasattr(fi, "fifty_two_week_low") else None,
            }
        except Exception as e:
            print(f"[getEtfDetails] Quote non disponible pour {ticker}: {e}")

        # Historical (1 an, interval 1j)
        historical = []
        try:
            hist_df = ticker_obj.history(period="1y", interval="1d")
            if not hist_df.empty:
                for idx, row in hist_df.iterrows():
                    historical.append({
                        "date": idx.isoformat() if hasattr(idx, "isoformat") else str(idx),
                        "open": float(row.get("Open", 0) or 0),
                        "high": float(row.get("High", 0) or 0),
                        "low": float(row.get("Low", 0) or 0),
                        "close": float(row.get("Close", 0) or 0),
                        "volume": int(row.get("Volume", 0) or 0),
                        "adjClose": float(row.get("Close", 0) or 0),
                    })
        except Exception as e:
            print(f"[getEtfDetails] Historical non disponible pour {ticker}: {e}")

        # Sustainability/ESG
        esg_score = None
        try:
            sust = ticker_obj.get_sustainability()
            if sust is not None and not sust.empty:
                esg_score = float(sust.loc["totalEsg", "Value"]) if "totalEsg" in sust.index else None
        except Exception:
            pass

        return {
            "quote": quote,
            "historical": historical,
            "esgScore": esg_score,
            "sustainability": None,
        }
    except Exception as e:
        print(f"[getEtfDetails] Erreur pour {ticker}: {e}")
        return {"quote": None, "historical": [], "esgScore": None, "sustainability": None}


def get_etf_performance(ticker: str, period: str = "1y") -> Optional[dict]:
    """Récupère la performance d'un ETF sur une période donnée."""
    period_map = {"1y": "1y", "6m": "6mo", "3m": "3mo", "1m": "1mo"}
    yf_period = period_map.get(period, "1y")
    try:
        hist_df = yf.Ticker(ticker).history(period=yf_period)
        if hist_df is None or len(hist_df) < 2:
            return None
        closes = hist_df["Close"].dropna().tolist()
        if len(closes) < 2:
            return None
        first_price = closes[0]
        last_price = closes[-1]
        performance = ((last_price - first_price) / first_price) * 100
        hist_list = [{"close": c} for c in closes]
        volatility = _calculate_volatility(hist_list)
        return {
            "performance": round(performance, 2),
            "volatility": round(volatility, 2),
        }
    except Exception as e:
        print(f"[getEtfPerformance] Erreur pour {ticker}: {e}")
        return None


def get_etf_holdings(ticker: str) -> Optional[dict]:
    """Récupère les holdings/composition d'un ETF."""
    try:
        ticker_obj = yf.Ticker(ticker)
        holders = ticker_obj.get_institutional_holders()
        if holders is not None and not holders.empty:
            return holders.to_dict(orient="records")
        return None
    except Exception as e:
        print(f"[getEtfHoldings] Erreur pour {ticker}: {e}")
        return None


def get_etfs(filters: dict = {}) -> list:
    """Récupère la liste des ETF avec données Yahoo Finance (ou fallback statique)."""
    if not TICKERS:
        return []

    quotes = []
    has_yahoo_error = False

    def _fetch_one(ticker: str):
        fi = yf.Ticker(ticker).fast_info
        last_price = fi.last_price if hasattr(fi, "last_price") and fi.last_price else None
        if last_price is None:
            return None
        prev_close = fi.previous_close if hasattr(fi, "previous_close") and fi.previous_close else None
        change = (last_price - prev_close) if prev_close else 0.0
        change_pct = (change / prev_close * 100) if prev_close else 0.0
        hi = fi.fifty_two_week_high if hasattr(fi, "fifty_two_week_high") else None
        lo = fi.fifty_two_week_low if hasattr(fi, "fifty_two_week_low") else None
        perf1y = round((hi - lo) / lo * 100, 2) if hi and lo and lo != 0 else 0.0
        volume = int(fi.three_month_average_volume) if hasattr(fi, "three_month_average_volume") and fi.three_month_average_volume else 0
        m = META.get(ticker, {})
        esg = m.get("esg", "A")
        ter = float(m.get("ter", 0.2))
        return {
            "id": ticker,
            "name": m.get("description", ticker),
            "ticker": ticker,
            "lastPrice": round(last_price, 2),
            "volume": volume,
            "change": round(change, 2),
            "changePercent": round(change_pct, 2),
            "ter": round(ter, 2),
            "perf1y": perf1y,
            "esg": esg,
            "match": _calc_match(ter, esg),
            "zone": m.get("zone", "Monde"),
            "theme": m.get("theme", "Diversifié"),
        }

    with ThreadPoolExecutor(max_workers=10) as pool:
        futures = {pool.submit(_fetch_one, t): t for t in TICKERS}
        for future in as_completed(futures):
            ticker = futures[future]
            try:
                result = future.result()
                if result:
                    quotes.append(result)
            except Exception as e:
                print(f"[getEtfs] Erreur pour {ticker}: {e}")

    if not quotes:
        has_yahoo_error = True

    # Fallback statique si Yahoo Finance indisponible
    if has_yahoo_error:
        print("[getEtfs] Fallback statique (Yahoo Finance indisponible)")
        for ticker in TICKERS:
            m = META.get(ticker, {})
            random_change = (random.random() - 0.5) * 2
            random_price = 100 + (random.random() - 0.5) * 20
            esg = m.get("esg", "A")
            quotes.append({
                "id": ticker,
                "name": m.get("description", ticker),
                "ticker": ticker,
                "lastPrice": round(random_price, 2),
                "volume": int(random.random() * 100000),
                "change": round(random_change, 2),
                "changePercent": round(random_change / random_price * 100, 2),
                "ter": 0.2,
                "perf1y": round((random.random() - 0.2) * 20, 2),
                "esg": esg,
                "match": _calc_match(0.2, esg),
                "zone": m.get("zone", "Monde"),
                "theme": m.get("theme", "Diversifié"),
                "isMock": True,
            })

    # Filtrage
    zone_filter = filters.get("zone")
    sector_filter = filters.get("sector")
    esg_filter = filters.get("esg")
    ter_max = filters.get("terMax")
    ESG_ORDER = ["B", "A", "AA", "AAA"]

    def _keep(etf: dict) -> bool:
        if zone_filter and zone_filter != "Toutes" and etf.get("zone") != zone_filter:
            return False
        if sector_filter and sector_filter != "Tous" and etf.get("theme") != sector_filter:
            return False
        if esg_filter and esg_filter != "Tous":
            idx_e = ESG_ORDER.index(etf["esg"]) if etf["esg"] in ESG_ORDER else -1
            idx_s = ESG_ORDER.index(esg_filter) if esg_filter in ESG_ORDER else -1
            if idx_e == -1 or idx_e < idx_s:
                return False
        if ter_max:
            import re as _re
            m = _re.search(r"0[,.](\d+)", ter_max)
            if m:
                max_val = float(f"0.{m.group(1)}")
                if etf.get("ter", 0) > max_val:
                    return False
        return True

    return [e for e in quotes if _keep(e)]
